spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.datasketches.theta.CompactSketch;
+import org.apache.datasketches.theta.SetOperationBuilder;
+import org.apache.datasketches.theta.UpdateSketch;
+import org.apache.iceberg.GenericBlobMetadata;
+import org.apache.iceberg.GenericStatisticsFile;
+import org.apache.iceberg.HasTableOperations;
+import org.apache.iceberg.StatisticsFile;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.puffin.Blob;
+import org.apache.iceberg.puffin.Puffin;
+import org.apache.iceberg.puffin.PuffinWriter;
+import org.apache.iceberg.puffin.StandardBlobTypes;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.function.Function2;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.internal.SQLConf;
+import scala.Tuple2;
+
+public class NDVSketchGenerator {
+
+  private NDVSketchGenerator() {}
+
+  static StatisticsFile generateNDV(
+      SparkSession spark, Table table, long snapshotId, String... columnsToBeAnalyzed)
+      throws IOException {
+    List<String> columnList;
+    if (columnsToBeAnalyzed == null || columnsToBeAnalyzed.length == 0) {
+      columnList =
+          table.schema().columns().stream()
+              .map(Types.NestedField::name)
+              .collect(Collectors.toList());
+    } else {
+      columnList = Lists.newArrayList(columnsToBeAnalyzed);
+    }
+    Iterator<Tuple2<String, ThetaSketchJavaSerializable>> tuple2Iterator =
+        computeNDVSketches(spark, table.name(), snapshotId, columnList);
+    Map<String, ThetaSketchJavaSerializable> sketchMap = Maps.newHashMap();
+
+    tuple2Iterator.forEachRemaining(tuple -> sketchMap.put(tuple._1, tuple._2));
+    return writeToPuffin(table, columnList, sketchMap);
+  }
+
+  private static StatisticsFile writeToPuffin(
+      Table table, List<String> columns, Map<String, ThetaSketchJavaSerializable> sketchMap)
+      throws IOException {
+    TableOperations operations = ((HasTableOperations) table).operations();
+    FileIO fileIO = operations.io();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = fileIO.newOutputFile(path);
+    try (PuffinWriter writer =
+        Puffin.write(outputFile).createdBy("Iceberg Analyze action").build()) {
+      for (String columnName : columns) {
+        writer.add(
+            new Blob(
```

Comment:
@findepi @marton-bod would be good to take a look as well, to verify interop between Spark and Trino here?

Here it seems we are storing the serialized sketch, as is specified in the spec.  Should we store ndv as well in 'metadata', as is specified in spec:  https://github.com/apache/iceberg/blob/main/format/puffin-spec.md#apache-datasketches-theta-v1-blob-type

Reply:
thanks @szehon-ho for the ping
there are couple potential issues

- the blob needs to have `ndv` property
- the sketch needs to be updated with the standard byte[] representation of values (`Conversions.toByteBuffer`)
- there should be inter-op test https://github.com/apache/iceberg/pull/10288#discussion_r1639569349
- i am not exactly sure how what's the lifecycle of `ThetaSketchJavaSerializable` & whether this can impact the final results. need to re-read this portion